### PR TITLE
Symmetric backup

### PR DESCRIPTION
**Pull request by ismailgulek**
Implements https://github.com/matrix-org/matrix-spec-proposals/pull/3270

Inspired by https://github.com/matrix-org/matrix-js-sdk/pull/1775

### CODE REVIEW

File: MatrixSDKTests/MXBaseKeyBackupTests.m [hunk 1]
```diff
@@ -628,21 +607,18 @@ - (void)testEncryptAndDecryptKeyBackupData
 
         // - Pick a megolm key
         MXOlmInboundGroupSession *session = [aliceSession.crypto.store inboundGroupSessionsToBackup:1].firstObject;
+        session.untrusted = self.isUntrusted;
```
Comment (manuroe) [inline MatrixSDKTests/MXBaseKeyBackupTests.m:610]: Why do we need that?
I would suggest to add a sanity check instead to make sure that all sessions handled outside of the key backup are trusted.

```
XCTAssertTrue(session.untrusted);
```

File: MatrixSDKTests/MXBaseKeyBackupTests.m [hunk 2]
```diff
@@ -628,21 +607,18 @@ - (void)testEncryptAndDecryptKeyBackupData
...
                 XCTAssertNotNil(keyBackupData.sessionData);
 
-                // - Check [MXKeyBackup pkDecryptionFromRecoveryKey] is able to create a OLMPkDecryption
-                OLMPkDecryption *decryption = [aliceSession.crypto.backup pkDecryptionFromRecoveryKey:keyBackupCreationInfo.recoveryKey error:nil];
-                XCTAssertNotNil(decryption);
-
-                // - Check [MXKeyBackup decryptKeyBackupData] returns stg
-                MXMegolmSessionData *sessionData = [aliceSession.crypto.backup decryptKeyBackupData:keyBackupData forSession:session.session.sessionIdentifier inRoom:roomId withPkDecryption:decryption];
+                // - Check [MXKeyBackupAlgorithm decryptKeyBackupData] returns stg
+                MXMegolmSessionData *sessionData = [aliceSession.crypto.backup.keyBackupAlgorithm decryptKeyBackupData:keyBackupData forSession:session.session.sessionIdentifier inRoom:roomId];
```
Comment (manuroe) [inline MatrixSDKTests/MXBaseKeyBackupTests.m:621]: To follow on the previous comment, we can have this test here:
```
   XCTAssertEqual(sessionData.untrusted , self.isUntrusted);
```

**Review by manuroe — COMMENTED** [review]
Thanks for this huge update.

My single concern is that the untrusted field is a bit fragile on a long term (even if we will remove this code with ElementR-iOS). A dev can change the code behavior and untrusted becomes YES even for keys received in a normal way with `m.room_key` to_device events. The resulting UX could be really bad.

I proposed a change in `testEncryptAndDecryptKeyBackupData` to verify that as a low cost way.

File: MatrixSDK/Crypto/KeyBackup/Data/Aes256/MXAes256KeyBackupAlgorithm.m [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+{
+    if (self = [super init])
+    {
+        if (authData == nil || ![authData isKindOfClass:MXAes256BackupAuthData.class])
+        {
+            MXLogError(@"[MXAes256KeyBackupAlgorithm] init: auth data missing required information");
+            return nil;
+        }
+        self.crypto = crypto;
+        NSParameterAssert([authData isKindOfClass:MXAes256BackupAuthData.class]);
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Crypto/KeyBackup/Data/Aes256/MXAes256KeyBackupAlgorithm.m:57]: It's totally useless now, removed.

File: MatrixSDK/Crypto/KeyBackup/MXKeyBackup.m [hunk 4]
```diff
@@ -405,10 +438,12 @@ - (void)restoreKeyBackupAutomaticallyWithPrivateKey:(void (^)(void))onComplete
...
     // Check private keys validity
-    NSString *pKDecryptionPublicKey = [self pkDecrytionPublicKeyFromCryptoStore];
-    if (![self checkPkDecryptionPublicKey:pKDecryptionPublicKey forKeyBackupVersion:self.keyBackupVersion])
+    NSData *privateKey = self.privateKeyFromCryptoStore;
+    NSError *error;
+    BOOL keyMatches = [_keyBackupAlgorithm keyMatches:privateKey error:&error];
+    if (error || !keyMatches)
     {
-        MXLogDebug(@"[MXKeyBackup] restoreKeyBackupAutomatically. Error: Invalid private key (%@)", pKDecryptionPublicKey);
+        MXLogDebug(@"[MXKeyBackup] restoreKeyBackupAutomatically. Error: Invalid private key (%@)", privateKey);
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Crypto/KeyBackup/MXKeyBackup.m:446]: Good catch, updated.

File: MatrixSDK/Crypto/KeyBackup/MXKeyBackup.m [hunk 5]
```diff
@@ -1029,38 +1052,28 @@ - (MXHTTPOperation*)restoreUsingPrivateKeyKeyBackup:(MXKeyBackupVersion*)keyBack
...
-        // Validate the local private key
-        if (decryption)
+
+        NSData *privateKey = self.privateKeyFromCryptoStore;
+        NSError *error;
+        if (error || ![[self getOrCreateKeyBackupAlgorithmFor:keyBackupVersion privateKey:privateKey] keyMatches:privateKey error:&error])
         {
-            NSString *pKDecryptionPublicKey = [self pkDecrytionPublicKeyFromCryptoStore];
-            if (![self checkPkDecryptionPublicKey:pKDecryptionPublicKey forKeyBackupVersion:keyBackupVersion])
+            MXLogDebug(@"[MXKeyBackup] restoreUsingPrivateKeyKeyBackup. Error: Invalid private key (%@) for %@", privateKey, keyBackupVersion);
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Crypto/KeyBackup/MXKeyBackup.m:1060]: Good catch, i've found an extra one and updated as well.

File: MatrixSDKTests/MXBaseKeyBackupTests.m (hunk 2 above)
Comment (ismailgulek) [reply MatrixSDKTests/MXBaseKeyBackupTests.m:622]: Good idea!

File: MatrixSDKTests/MXBaseKeyBackupTests.m (hunk 1 above)
Comment (ismailgulek) [reply MatrixSDKTests/MXBaseKeyBackupTests.m:611]: Yeah good idea, have added that. But we still need to alter the flag to have the same untrusted value when comparing `sessionData` objects later.

**Comment by ismailgulek** [comment]
I've checked the failing integration & crypto tests, they were green on my local.

**Review by manuroe — APPROVED** [review]
LGTM!

**Pull request merged**